modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/SourceHandler.java

```diff
@@ -154,6 +166,20 @@ public void requestReceived(NHttpServerConnection conn) {
         }
     }
 
+    private void setCorrelationId(HttpContext httpContext, SourceRequest request) {
+
+        String externalCorrelationId = request.getHeaders().get(PassThroughConstants.CORRELATION_ID);
+        String corId;
+        long corTime = System.currentTimeMillis();
+        if (externalCorrelationId == null) {
+            corId = UUID.randomUUID().toString();
+        } else {
+            corId = externalCorrelationId;
+        }
+        httpContext.setAttribute(PassThroughConstants.CORRELATION_ID, corId);
+        httpContext.setAttribute(PassThroughConstants.CORRELATION_REQUEST_ARRIVED_TIME, corTime);
```

Comment:
method name says we set the correlationId here, so I think better not to set the request arrival time here.
We can set that property from the place where we are calling this method. Since this method need request arrival time, we can parse it as a parameter in the method.